From 62932c7e33c7919ea8160a2b6660df501c7507fe Mon Sep 17 00:00:00 2001
From: Frank Wang <frank.wang@rock-chips.com>
Date: Mon, 22 Jan 2018 15:02:49 +0800
Subject: [PATCH] usb: amend ehci no-relinquish-port for rk3288 platform

For the hardware bug of RK3288 OHCI, we use commit cfe6f1dd57e0
("usb: ehci: add rockchip relinquishing port quirk support") to fix
it previously. However, it have been ineffective after upstream commit
94c43b9897ab ("USB: Check for dropped connection before switching to
full speed") was merged due to the condition of relinquishing port was
changed.

This patch adds an additional condition for the previous commit to ensure
no relinquish port quirk can take effect for RK3288 EHCI.

Change-Id: I0630265e101afb349816955e069e1c121745ac08
Signed-off-by: Frank Wang <frank.wang@rock-chips.com>
---
 drivers/usb/core/hub.c           | 3 ++-
 drivers/usb/host/ehci-platform.c | 4 +++-
 include/linux/usb/hcd.h          | 1 +
 3 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 77ab0d4fff44..097a2ac9c2bb 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -4871,7 +4871,8 @@ static void hub_port_connect(struct usb_hub *hub, int port1, u16 portstatus,
 done:
 	hub_port_disable(hub, port1, 1);
 	if (hcd->driver->relinquish_port && !hub->hdev->parent) {
-		if (status != -ENOTCONN && status != -ENODEV)
+		if ((status != -ENOTCONN && status != -ENODEV) ||
+		    (status == -ENOTCONN && hcd->rk3288_relinquish_port_quirk))
 			hcd->driver->relinquish_port(hcd, port1);
 	}
 }
diff --git a/drivers/usb/host/ehci-platform.c b/drivers/usb/host/ehci-platform.c
index bedf6565caa4..70e46ec73440 100644
--- a/drivers/usb/host/ehci-platform.c
+++ b/drivers/usb/host/ehci-platform.c
@@ -216,9 +216,11 @@ static int ehci_platform_probe(struct platform_device *dev)
 
 		if (of_machine_is_compatible("rockchip,rk3288") &&
 		    of_property_read_bool(dev->dev.of_node,
-					  "rockchip-relinquish-port"))
+					  "rockchip-relinquish-port")) {
 			ehci_platform_hc_driver.relinquish_port =
 					  ehci_rockchip_relinquish_port;
+			hcd->rk3288_relinquish_port_quirk = 1;
+		}
 
 		priv->num_phys = of_count_phandle_with_args(dev->dev.of_node,
 				"phys", "#phy-cells");
diff --git a/include/linux/usb/hcd.h b/include/linux/usb/hcd.h
index 757c554408ce..acce195be773 100644
--- a/include/linux/usb/hcd.h
+++ b/include/linux/usb/hcd.h
@@ -159,6 +159,7 @@ struct usb_hcd {
 	unsigned		tpl_support:1; /* OTG & EH TPL support */
 	unsigned		cant_recv_wakeups:1;
 			/* wakeup requests from downstream aren't received */
+	unsigned		rk3288_relinquish_port_quirk:1;
 
 	unsigned int		irq;		/* irq allocated */
 	void __iomem		*regs;		/* device memory/io */
-- 
2.35.3

